### PR TITLE
WIP: Flex Microform Spike

### PR DESCRIPTION
**Pull request by mulby**

### CODE REVIEW

File: ecommerce/extensions/basket/views.py [hunk 1]
```diff
@@ -669,6 +669,7 @@ def _serialize_context(self, context, lines_data):
         self._add_total_summary(response, context)
         self._add_offers(response)
         self._add_coupons(response, context)
+        self._add_capture_context(response)
```
Comment (mulby) [inline ecommerce/extensions/basket/views.py:672]: change "capture context" to "public key"

File: ecommerce/extensions/basket/views.py [hunk 2]
```diff
@@ -724,6 +725,19 @@ def _add_coupons(self, response, context):
...
     def _add_messages(self, response):
         response['messages'] = message_utils.serialize(self.request)
+    
+    def _add_capture_context(self, response):
+        response['flex_microform_enabled'] = waffle.flag_is_active(self.request, 'payment.cybersource.flex_microform_enabled')
+        if not response['flex_microform_enabled']:
+            return
+        payment_processor_class = self.request.site.siteconfiguration.get_client_side_payment_processor_class()
+        if not payment_processor_class:
+            return
```
Comment (mulby) [inline ecommerce/extensions/basket/views.py:738]: This should fail more obviously

File: ecommerce/extensions/payment/processors/cybersource.py [hunk 3]
```diff
@@ -9,8 +9,17 @@
 from decimal import Decimal
 
 import six
+from CyberSource import (
```
Comment (mulby) [inline ecommerce/extensions/payment/processors/cybersource.py:14]: Requires python >=3.6

File: ecommerce/extensions/payment/processors/cybersource.py [hunk 4]
```diff
@@ -109,6 +131,169 @@ def cancel_page_url(self):
...
+
+        requestObj = GeneratePublicKeyRequest(
+            encryption_type='RsaOaep256',
+            target_origin=self.flex_target_origin,
+        )
+        requestObj = self.del_none(requestObj.__dict__)
+        requestObj = json.dumps(requestObj)
+
+        api_instance = KeyGenerationApi(self.cybersource_api_config)
+        return_data, status, body = api_instance.generate_public_key(requestObj, format='JWT')
```
Comment (mulby) [inline ecommerce/extensions/payment/processors/cybersource.py]: Timeouts and such!?!

File: ecommerce/extensions/payment/processors/cybersource.py [hunk 5]
```diff
@@ -109,6 +131,169 @@ def cancel_page_url(self):
     def client_side_payment_url(self):
         return self.sop_payment_page_url
 
+    def del_none(self, d):
```
Comment (inventhouse) [inline ecommerce/extensions/payment/processors/cybersource.py]: hack: wacky stuff from cybersource code example

File: ecommerce/extensions/payment/views/cybersource.py [hunk 6]
```diff
@@ -171,6 +171,88 @@ def post(self, request):
...
+        logger.info(
+            '%s called for basket [%d]. It is in the [%s] state.',
+            self.__class__.__name__,
+            request.basket.id,
+            request.basket.status
+        )
+        return super(CybersourceAuthorizeAPIView, self).post(request)
+
+    def form_valid(self, form):
+        data = form.cleaned_data
```
Comment (mulby) [inline ecommerce/extensions/payment/views/cybersource.py]: A bunch of shared code between this and the old "api-submit" endpoint

File: ecommerce/extensions/payment/views/cybersource.py [hunk 7]
```diff
@@ -171,6 +171,88 @@ def post(self, request):
...
+            return JsonResponse(response_to_return, status=403)
+
+        logger.info(
+            'SDNCheck function called for basket [%d]. It did not receive a hit.',
+            request.basket.id,
+        )
+
+        try:
+            handled_processor_response = self.payment_processor.authorize_payment(basket, request, data)
+        except GatewayError:
```
Comment (mulby) [inline ecommerce/extensions/payment/views/cybersource.py]: Check all this error handling very carefully

File: ecommerce/extensions/payment/views/cybersource.py [hunk 8]
```diff
@@ -171,6 +171,88 @@ def post(self, request):
...
+            line2=data['address_line2'],
+            line4=data['city'],
+            postcode=data['postal_code'],
+            state=data['state'],
+            country=Country.objects.get(iso_3166_1_a2=data['country'])
+        )
+
+        with transaction.atomic():
+            basket.freeze()
+            self.record_payment(basket, handled_processor_response)
```
Comment (mulby) [inline ecommerce/extensions/payment/views/cybersource.py]: This is the core of the paper trail risk

File: ecommerce/extensions/payment/views/cybersource.py [hunk 9]
```diff
@@ -171,6 +171,88 @@ def post(self, request):
...
+            request.site.siteconfiguration,
+            order_number=basket.order_number,
+            disable_back_button=True,
+        )
+        return JsonResponse({
+            'receipt_page_url': receipt_page_url,
+        }, status=201)
+
+
 class CybersourceInterstitialView(CyberSourceProcessorMixin, EdxOrderPlacementMixin, View):
```
Comment (mulby) [inline ecommerce/extensions/payment/views/cybersource.py]: Should we refactor the core of this to extract the data gathering from the request and the processing?

**Review by MatthewPiatetsky — APPROVED** [review]
This code will needs tests, cleanup, documentation and so on, but we are merging a draft version of the code for now to keep things moving

**Pull request merged**